subsys/trusted_storage/src/auth_crypt_backend.c

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+#include <zephyr/device.h>
+#include <init.h>
+#include <sys/util.h>
+
+#include <stdio.h>
+#include <stdint.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+
+#include "settings_helpers.h"
+#include "auth_crypt_backend.h"
+
+/* AEAD IV suffix */
+#define TRUSTED_STORAGE_FILENAME_SUFFIX_IV		".iv"
+
+/*
+ * AEAD based Authenticated Encrypted trust implementation
+ *
+ * Actual implementation uses:
+ * - Hexadecimal UID imported as KEY
+ * - UID+Flags+Size as additional parameter
+ * - IV is an incrementing number at each encryption
+ * - TAG is left at the end of output data
+ */
+
+/* TODO Kconfig */
+#define AEAD_TAG_SIZE		16
+#define AEAD_IV_SIZE		12
+
+/* Max storage size for the encrypted or decrypted output */
+#define AEAD_MAX_BUF_SIZE ROUND_UP(TRUSTED_STORAGE_MAX_ASSET_SIZE +	\
+				   AEAD_TAG_SIZE, AEAD_TAG_SIZE)
+
+/* 128bits AEAD key pattern: uid low, uid */
+#define AEAD_KEY_PATTERN	"%08x%08x"
+#define AEAD_KEY_SIZE		(sizeof(uint64_t) * 2)
+
+/* Additional data structure */
+struct aead_additional_data {
+	psa_storage_uid_t uid;
+	psa_storage_create_flags_t flags;
+	size_t size;
+};
+
+static uint64_t g_iv_low;
+static uint64_t g_iv_high;
+
+/* Return an incrementing IV */
+static psa_status_t psa_ps_get_iv(uint8_t *iv, size_t iv_len)
```

Comment:
I would appreciate a clearly marked API to handle initial vector and making a counter scheme internal to this. The fact that this is a static function and it has extra functionality beyond just getting the IV (i.e. implementing a counter scheme on some of the IV data) is 

I would also recommend hiding the storage of the actual IV beneath a clearly marked API in case the system supports running counters or externally sourced IV from hw peripherals